pkg/receiver/rawk8seventsreceiver/receiver.go

```diff
@@ -144,6 +160,64 @@ func (r *rawK8sEventsReceiver) Shutdown(context.Context) error {
 	return nil
 }
 
+func (r *rawK8sEventsReceiver) getStorage(ctx context.Context, host component.Host) (storage.Client, error) {
+	if host == nil {
+		r.logger.Debug("Storage not initialized: host is not available")
+		return nil, nil
+	}
+
+	var storageExtension storage.Extension
+	var storageExtensionId config.ComponentID
+	for extentionId, extension := range host.GetExtensions() {
+		if se, ok := extension.(storage.Extension); ok {
+			if storageExtension != nil {
+				return nil, fmt.Errorf("multiple storage extensions found: '%s', '%s'", storageExtensionId, extentionId)
+			}
+			storageExtension = se
+			storageExtensionId = extentionId
+		}
+	}
+
+	if storageExtension == nil {
+		r.logger.Debug("Storage not initialized: no storage extension found")
+		return nil, nil
+	}
+
+	storageClient, err := storageExtension.GetClient(ctx, component.KindReceiver, r.cfg.ID(), "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get storage client for extension '%s': %s", storageExtensionId, err)
+	}
+
+	r.logger.Info("Initialized storage", zap.Any("storage_extension_id", storageExtensionId))
+	return storageClient, nil
+}
+
+func (r *rawK8sEventsReceiver) getStartResourceVersion(ctx context.Context) (uint64, error) {
+	if r.storage == nil {
+		r.logger.Info("Did not find start resource version, as there is no storage.")
+		return 0, nil
+	}
+
+	startResourceVersionBytes, err := r.storage.Get(ctx, "startResourceVersion")
+	if err != nil {
+		return 0, fmt.Errorf("failed to retrieve start resource version from storage: %s", err)
+	}
+
+	if startResourceVersionBytes == nil {
+		r.logger.Info("Start resource version not found in storage")
+		return 0, nil
+	}
+
+	startResourceVersionString := string(startResourceVersionBytes)
+	startResourceVersion, err := strconv.ParseUint(startResourceVersionString, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse start resource version '%s' to number: %s", startResourceVersionString, err)
+	}
```

Comment:
Looks kind of complicated for reading an int from a KV store. Maybe use https://pkg.go.dev/encoding/binary instead? Probably faster as well.

Reply:
Using the `binary` package doesn't make things better, in fact I would argue it makes things worse as there is even more boilerplate code:

```patch
---
 pkg/receiver/rawk8seventsreceiver/receiver.go | 24 +++++++++++++++----
 1 file changed, 19 insertions(+), 5 deletions(-)

diff --git a/pkg/receiver/rawk8seventsreceiver/receiver.go b/pkg/receiver/rawk8seventsreceiver/receiver.go
index 43cfbbdb..8ed54e16 100644
--- a/pkg/receiver/rawk8seventsreceiver/receiver.go
+++ b/pkg/receiver/rawk8seventsreceiver/receiver.go
@@ -16,6 +16,7 @@ package rawk8seventsreceiver
 
 import (
 	"context"
+	"encoding/binary"
 	"errors"
 	"fmt"
 	"strconv"
@@ -214,10 +215,10 @@ func (r *rawK8sEventsReceiver) getLatestResourceVersion(ctx context.Context) (ui
 		return 0, nil
 	}
 
-	latestResourceVersionString := string(latestResourceVersionBytes)
-	latestResourceVersion, err := strconv.ParseUint(latestResourceVersionString, 10, 64)
-	if err != nil {
-		return 0, fmt.Errorf("failed to parse latest resource version '%s' to number: %s", latestResourceVersionString, err)
+	latestResourceVersion, length := binary.Uvarint(latestResourceVersionBytes)
+	if length <= 0 {
+		r.logger.Warn("failed to parse latest resource version", zap.Any("length", length))
+		return 0, nil
 	}
 
 	r.logger.Info("Found latest resource version in storage", zap.Any("latest_resource_version", latestResourceVersion))
@@ -293,7 +294,20 @@ func (r *rawK8sEventsReceiver) recordEventProcessed(event *corev1.Event) {
 		return
 	}
 
-	r.storage.Set(r.ctx, latestResourceVersionStorageKey, []byte(event.ResourceVersion))
+	resourceVersionNumber, err := strconv.ParseUint(event.ResourceVersion, 10, 64)
+	if err != nil {
+		r.logger.Debug("Cannot record incoming resource version, not a number.", zap.Any("incoming_resource_version", event.ResourceVersion))
+		return
+	}
+
+	bytes := make([]byte, binary.MaxVarintLen64)
+	length := binary.PutUvarint(bytes, resourceVersionNumber)
+	if length <= 0 {
+		r.logger.Warn("Cannot record incoming resource version, cannot encode.", zap.Any("length", length), zap.Any("incoming_resource_version", event.ResourceVersion))
+		return
+	}
+
+	r.storage.Set(r.ctx, latestResourceVersionStorageKey, bytes[:length])
 }
 
 // Check if we should process the event.
```